# Drawing Conclusions
Use the space below to address questions on datasets `clean_08.csv` and `clean_18.csv`. You should've created these data files in the previous section: *Fixing Data Types Pt 3*.

In [190]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = 'all'

%config IPCompleter.greedy = True

In [191]:
import pandas as pd
import numpy as np
%matplotlib inline

In [192]:
# load 2008 dataset
df_08.shape
df_08 = pd.read_csv('clean_08.csv')
df_08.head(n = 2)

(987, 13)

model  displ  cyl    trans drive      fuel veh_class  \
0  ACURA MDX    3.7  6.0  Auto-S5   4WD  Gasoline       SUV   
1  ACURA RDX    2.3  4.0  Auto-S5   4WD  Gasoline       SUV   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                  7.0      15.0     20.0     17.0                   4.0   
1                  7.0      17.0     22.0     19.0                   5.0   

  smartway  
0       no  
1       no

In [193]:
# load 2018 dataset
df_18 = pd.read_csv('clean_18.csv')
df_18.shape
df_18.head(n = 2)

(832, 13)

model  displ  cyl       trans drive      fuel  veh_class  \
0  ACURA RDX    3.5  6.0  SemiAuto-6   2WD  Gasoline  small SUV   
1  ACURA RDX    3.5  6.0  SemiAuto-6   4WD  Gasoline  small SUV   

   air_pollution_score  city_mpg  hwy_mpg  cmb_mpg  greenhouse_gas_score  \
0                  3.0      20.0     28.0     23.0                   5.0   
1                  3.0      19.0     27.0     22.0                   4.0   

  smartway  
0       No  
1       No

### Q1: Are more unique models using alternative sources of fuel? By how much?

In [194]:
df_08.fuel.value_counts()

Gasoline    984
ethanol       1
CNG           1
gas           1
Name: fuel, dtype: int64

In [195]:
unique_fuel_source_count = df_18.fuel.value_counts()
total_fuel_source_count = df_18.fuel.value_counts().sum()

differ = 100- (unique_fuel_source_count[0]/total_fuel_source_count)*100
differ

9.975961538461547

> In 2018 9.98 % alternative source of fuel is used by unique models


### Q2: How much have vehicle classes improved in fuel economy?  

In [196]:
df_08.columns

Index(['model', 'displ', 'cyl', 'trans', 'drive', 'fuel', 'veh_class',
       'air_pollution_score', 'city_mpg', 'hwy_mpg', 'cmb_mpg',
       'greenhouse_gas_score', 'smartway'],
      dtype='object')

##### Mean of cmb_mpg based on various veh_class in 2008 dataset

In [197]:
veh_class_grp_08_df = df_08.groupby('veh_class')
veh_class_grp_08_cmb_mpg_mean_series = veh_class_grp_08_df['cmb_mpg'].mean()

##### Mean of cmb_mpg based on various veh_class in 2018 dataset

In [198]:
veh_class_grp_18_df = df_18.groupby('veh_class')
veh_class_grp_18_cmb_mpg_mean_series = veh_class_grp_18_df['cmb_mpg'].mean()

In [199]:
#### Increment in cmb_mpg

In [200]:
difference = veh_class_grp_18_cmb_mpg_mean_series - veh_class_grp_08_cmb_mpg_mean_series
difference

veh_class
SUV                     NaN
large car          4.900000
midsize car        6.282609
minivan            1.682353
pickup             2.312635
small SUV               NaN
small car          4.315948
special purpose         NaN
standard SUV            NaN
station wagon      5.162745
van                     NaN
Name: cmb_mpg, dtype: float64

### Q3: What are the characteristics of SmartWay vehicles? Have they changed over time?

In [201]:
df_08.columns

Index(['model', 'displ', 'cyl', 'trans', 'drive', 'fuel', 'veh_class',
       'air_pollution_score', 'city_mpg', 'hwy_mpg', 'cmb_mpg',
       'greenhouse_gas_score', 'smartway'],
      dtype='object')

In [202]:
smartway_unique_values_df_08 = df_08.loc[:, 'smartway'].unique()
smartway_unique_values_df_08

array(['no', 'yes'], dtype=object)

In [203]:
# filter df_08 datasets, only for smartway with yes value and get statisctics for each column
df_08[df_08['smartway'] == smartway_unique_values_df_08[1]].describe()

displ         cyl  air_pollution_score    city_mpg     hwy_mpg  \
count  380.000000  380.000000           380.000000  380.000000  380.000000   
mean     2.602895    4.826316             7.365789   20.984211   28.413158   
std      0.623436    1.002025             1.148195    3.442672    3.075194   
min      1.300000    4.000000             6.000000   17.000000   22.000000   
25%      2.275000    4.000000             7.000000   19.000000   26.000000   
50%      2.400000    4.000000             7.000000   20.000000   28.000000   
75%      3.000000    6.000000             7.000000   22.000000   30.000000   
max      5.000000    8.000000             9.500000   48.000000   45.000000   

          cmb_mpg  greenhouse_gas_score  
count  380.000000            380.000000  
mean    23.736842              6.868421  
std      3.060379              0.827338  
min     20.000000              6.000000  
25%     22.000000              6.000000  
50%     23.000000              7.000000  
75%     25.000000              7.000000  
max     46.000000             10.000000

In [204]:
smartway_unique_values_df_18 = df_18.loc[:, 'smartway'].unique()
smartway_unique_values_df_18

array(['No', 'Yes', 'Elite'], dtype=object)

In [205]:
# filter df_18 datasets, only for smartway with yes value and get statisctics for each column
df_18[df_18['smartway'] == smartway_unique_values_df_18[1]].describe()

displ        cyl  air_pollution_score    city_mpg    hwy_mpg  \
count  91.000000  91.000000            91.000000   91.000000  91.000000   
mean    1.753846   3.923077             4.879121   31.483516  39.296703   
std     0.420399   0.452911             1.769087   11.239180   9.410980   
min     1.200000   3.000000             3.000000   25.000000  27.000000   
25%     1.500000   4.000000             3.000000   28.000000  36.000000   
50%     1.600000   4.000000             5.000000   28.000000  37.000000   
75%     2.000000   4.000000             7.000000   30.000000  40.000000   
max     3.500000   6.000000             7.000000  113.000000  99.000000   

          cmb_mpg  greenhouse_gas_score  
count   91.000000             91.000000  
mean    34.439560              7.538462  
std     10.214596              0.860332  
min     27.000000              7.000000  
25%     31.000000              7.000000  
50%     32.000000              7.000000  
75%     34.000000              8.000000  
max    106.000000             10.000000

### Q4: What features are associated with better fuel economy?

In [206]:
cmb_mpg_08_mean = df_08.loc[:, 'cmb_mpg'].mean()
cmb_mpg_08_mean

cmb_mpg_above_mean_08_df = df_08.query('cmb_mpg > @cmb_mpg_08_mean')
cmb_mpg_above_mean_08_df

19.78824721377913